fix: NetworkSceneManager does not synchronize despawned in-scene placed NetworkObjects [MTT-2924]

NoelStephensUnity opened the pull request:
This resolves the issue where an in-scene placed NetworkObject could be despawned prior to a client joining and the client-side in-scene placed NetworkObject component's would not receive a despawn notification when synchronizing.

This also greatly reduces user pain by not requiring users to register in-scene placed NetworkObjets with the NetworkManager using prefab hash overrides.  Instead, it just uses the table of in-scene NetworkObjects that were already being tracked by the NetworkSceneManager.

[MTT-2924](https://jira.unity3d.com/browse/MTT-2924)

## Changelog
- Fixed: `NetworkSceneManager` does not synchronize despawned in-scene placed NetworkObjects.
- Changed: requirement to register in-scene placed NetworkObjects with `NetworkManager` in order to respawn them.  In-scene placed NetworkObjects are now automatically tracked during runtime and no longer need to be registered as a NetworkPrefab.

## Testing and Documentation
- Includes manual test (Additive Scene Loading: Backspace on server spawns/despawns the green in-scene NetworkObject)
- Includes integration test updates.
- Includes integration test: InScenePlacedNetworkObjects

## NetcodeIntegrationTest Updates:
![image](https://user-images.githubusercontent.com/73188597/165867537-b1019b31-9270-4647-834b-514a0d748c68.png)
The image above is an integration test that has two (2) clients and (1) host where the host loaded a scene that has in-scene placed NetworkObjects.  If the in-scene NetworkObject has the TestHelper ObjectNameIdentifier component, then like in the above image you will be able to visually distinguish between each relative instance.

NoelStephensUnity committed:
fix

MTT-2924
This is the first pass at solving the issue where late joining clients are not synchronized properly with despawned in-scene NetworkObjects.

NoelStephensUnity committed:
Update CHANGELOG.md

NoelStephensUnity committed:
Update CHANGELOG.md

NoelStephensUnity committed:
fix

This allows for respawning of in-scene placed NetworkObjects without having to register them with the NetworkManager.

NoelStephensUnity committed:
Update CHANGELOG.md

NoelStephensUnity committed:
style

NoelStephensUnity committed:
Merge branch 'develop' into fix/mtt-2924-despawned-scene-objects-dont-synchronize-late-joining-clients

ShadauxCat commented on com.unity.netcode.gameobjects/CHANGELOG.md:
```diff
@@ -14,7 +14,7 @@ Additional documentation and release notes are available at [Multiplayer Documen
 
 ### Fixed
-
+- Fixed `NetworkSceneManager` does not synchronize despawned in-scene placed NetworkObjects nor do they have to register them to re-spawn them. (#1898)
```
Grammar is a little weird here. Maybe "`NetworkSceneManager` no longer synchronizes..."

NoelStephensUnity replied:
Actually, I must have been sleepy when I wrote that.  :)
It is a fix and a change.

ShadauxCat commented on com.unity.netcode.gameobjects/Runtime/Spawning/NetworkSpawnManager.cs:
```diff
@@ -385,7 +385,7 @@ internal void ChangeOwnership(NetworkObject networkObject, ulong clientId)
         /// Should only run on the client
         /// </summary>
-        internal NetworkObject CreateLocalNetworkObject(bool isSceneObject, uint globalObjectIdHash, ulong ownerClientId, ulong? parentNetworkId, Vector3? position, Quaternion? rotation, bool isReparented = false)
+        internal NetworkObject CreateLocalNetworkObject(bool isSceneObject, uint globalObjectIdHash, ulong ownerClientId, ulong? parentNetworkId, int? networkSceneHandle, Vector3? position, Quaternion? rotation, bool isReparented = false)
```
Small nit, but I wonder what the pros/cons are of this vs just taking an int and assigning `NO_SCENE_HANDLE = -1`. The latter might perform a little better... but probably immeasurably so. So I'm just wondering which one is more elegant.

NoelStephensUnity replied:
I think -1 could be a valid scene handle... haven't seen anything like 0...
But yeah... most definitely this was something I was noodling over for awhile...my final choice was to follow the already existing pattern in this method where there just might not be a value available and so you pass "null" or an unassigned parameter of the specific type.

It was mainly to just keep the same usage pattern in the methods parameters...

ShadauxCat replied:
Ah. Well then. Nevermind me. *disappears*

NoelStephensUnity committed:
Update CHANGELOG.md

NoelStephensUnity committed:
style

whitespace fixes.

NoelStephensUnity committed:
Merge branch 'develop' into fix/mtt-2924-despawned-scene-objects-dont-synchronize-late-joining-clients

NoelStephensUnity committed:
Merge branch 'develop' into fix/mtt-2924-despawned-scene-objects-dont-synchronize-late-joining-clients

NoelStephensUnity committed:
Merge branch 'develop' into fix/mtt-2924-despawned-scene-objects-dont-synchronize-late-joining-clients

NoelStephensUnity committed:
fix

Passing the entire SceneObject structure as opposed to adding more parameters to the NetworkSpawnManager.HasPrefab method in order to be able to extract the NetworkSceneHandle and include the in-scene placed NetworkObjects as part of the "prefabs to search for".

NoelStephensUnity committed:
update

This is some work towards integration tests where clients might be able to actually load scenes as opposed to fake loading them

NoelStephensUnity committed:
fix

Fixed some minor issues with tests.
Updated HiddenVariableTests as it needed some timing related adjustments.

NoelStephensUnity committed:
Merge branch 'develop' into fix/mtt-2924-despawned-scene-objects-dont-synchronize-late-joining-clients

NoelStephensUnity committed:
Update NetworkSceneManagerEventDataPoolTest.cs

NoelStephensUnity committed:
style

NoelStephensUnity committed:
Merge branch 'develop' into fix/mtt-2924-despawned-scene-objects-dont-synchronize-late-joining-clients

NoelStephensUnity committed:
style

NoelStephensUnity committed:
test - integration test clients now load scenes

This update allows for clients to load scenes and synchronize in-scene placed NetworkObjects during integration testing.

NoelStephensUnity committed:
test - fix

Now that clients are loading scenes, needed to make some minor adjustments.

NoelStephensUnity committed:
test - fix

Realized I don't need to simulate client loading delay unless the client is blocked from loading.
Also had to adjust whether the metrics tests waited for the client or not when loading.

NoelStephensUnity committed:
Merge branch 'develop' into fix/mtt-2924-despawned-scene-objects-dont-synchronize-late-joining-clients

NoelStephensUnity committed:
fix

Fixes an issue that can occur where this becomes enabled in test runner and will throw an exception if target is null.

NoelStephensUnity committed:
fix - NetworkSceneManager.GetSceneRelativeInSceneNetworkObject

GetSceneRelativeInSceneNetworkObject needed to get the sceneHandle before checking for it existing in the ScenePlacedObjects.

NoelStephensUnity committed:
fix - AddDespawnedInSceneNetworkObjects

AddDespawnedInSceneNetworkObjects needed to only return despawned NetworkObjects belonging to the NetworkManager.

NoelStephensUnity committed:
test - fix - Integration test client scene loading

This fixes some of the last edge case scenarios such as:
- Not loading scenes for clients when a non-NetcodeIntegrationTest based integration test is running since this does require some additional management and cleanup.
- Assuring that in-scene NetworkObjects have the right NetworkManager assigned when loaded
- Added a check to assure that it didn't time out while waiting for all clients to connect
- Increased the NetcodeIntegrationTestHelpers default timeout to 2s vs 1s (clients loading scenes can slow down when things happen)
- NetcodeIntegrationTests now unload all scenes loaded (but the test runner scene) during teardown and onetimeteardown.

NoelStephensUnity committed:
test update

Just adding timeout checks in both NetworkVariableTests and TimeMultiInstanceTest

NoelStephensUnity committed:
test assets

Updating some of the assets used for integration testing now that clients are loading scenes too.

NoelStephensUnity committed:
test fix

Changing the scenes being loaded to avoid integration test specific SDK architecture related issues.

NoelStephensUnity committed:
test fix

An early shutdown was not needed here

NoelStephensUnity committed:
test fix

Disabling the audio listener in the EmptyScene to reduce console spam during integration tests.

NoelStephensUnity committed:
test fix

Fixing edge case timing issue with MessageOrdering.

NoelStephensUnity committed:
test InScenePlacedNetworkObjects

This is the integration test for this PR.
It tests late joining client synchronization with despawned NetworkObjects and tests that in-scene placed NetworkObjects do not need to be registered with the NetworkManager to be spawned.

NoelStephensUnity committed:
fix

Don't throw an exception when calling GetObservers or IsNetworkVisibleTo when the NetworkObject is not spawned.

NoelStephensUnity committed:
Merge branch 'develop' into fix/mtt-2924-despawned-scene-objects-dont-synchronize-late-joining-clients

NoelStephensUnity committed:
fix

fixing typo/mistake for the scene to be unloaded.

NoelStephensUnity committed:
style

NoelStephensUnity committed:
Update ObjectNameIdentifier.cs

NoelStephensUnity committed:
style

NoelStephensUnity committed:
Merge branch 'develop' into fix/mtt-2924-despawned-scene-objects-dont-synchronize-late-joining-clients

NoelStephensUnity committed:
style

updating comments and removing LFs and some commented out code.

NoelStephensUnity committed:
update and style

updating comments and minor final tweaks.

jeffreyrainy commented on com.unity.netcode.gameobjects/Runtime/Core/NetworkObject.cs:
```diff
@@ -369,7 +370,7 @@ private void SpawnInternal(bool destroyWithScene, ulong ownerClientId, bool play
             }
 
-            NetworkManager.SpawnManager.SpawnNetworkObjectLocally(this, NetworkManager.SpawnManager.GetNetworkObjectId(), false, playerObject, ownerClientId, destroyWithScene);
+            NetworkManager.SpawnManager.SpawnNetworkObjectLocally(this, NetworkManager.SpawnManager.GetNetworkObjectId(), IsSceneObject.HasValue && IsSceneObject.Value == true, playerObject, ownerClientId, destroyWithScene);
```
Is the `== true` really needed? If it's boolean, the `&&` should work already, or is there some funky casting going on?

NoelStephensUnity replied:
This was probably just a mistake on my part... will remove the == true as it is not needed at all.

jeffreyrainy commented on com.unity.netcode.gameobjects/Runtime/Core/NetworkObject.cs:
```diff
@@ -946,7 +961,7 @@ internal SceneObject GetMessageSceneObject(ulong targetClientId)
                     NetworkObjectId = NetworkObjectId,
                     OwnerClientId = OwnerClientId,
-                    IsSceneObject = IsSceneObject ?? true,
+                    IsSceneObject = IsSceneObject != false,
```
Wait, what? what's the difference between `IsSceneObject ?? true` and `IsSceneObject != false` ?

NoelStephensUnity replied:
Nothing... I could change that back...it was just easier for me at the time to see IsSceneObject != false than IsSceneObject ?? true.

NoelStephensUnity committed:
update

minor changes suggested by Jeffrey R.

NoelStephensUnity committed:
update

removing scene management specific code from places where scene management is disabled.

NoelStephensUnity committed:
style

updated some comments

NoelStephensUnity committed:
fix

Fixing an issue where CheckValueOnClient was crashing.  Integration test timing did change slightly now that clients load scenes, so it was possible that the key would not exist by the time it was being checked.

NoelStephensUnity committed:
fix  consoles

reducing the number of clients from 9 to 4 in some of the NetworkSceneManager related tests as I think it impacted performance enough to cause clients to take longer than the default timeout of 2 seconds.

NoelStephensUnity committed:
fix

Hopefully this will fix the issue with Xbox Scarlett

0xFA11 commented on com.unity.netcode.gameobjects/TestHelpers/Runtime/NetcodeIntegrationTestHelpers.cs:
```diff
@@ -16,7 +16,7 @@ namespace Unity.Netcode.TestHelpers.Runtime
     {
         public const int DefaultMinFrames = 1;
-        public const float DefaultTimeout = 1f;
+        public const float DefaultTimeout = 2f;
```
lol

NoelStephensUnity replied:
Well, there is additional overhead with loading scenes and adding that to the switch over to using only UTP...let's just say the timing has changed enough to increase the margin for error... :)

0xFA11 commented on com.unity.netcode.gameobjects/CHANGELOG.md:
```diff
@@ -18,7 +18,7 @@ Additional documentation and release notes are available at [Multiplayer Documen
 
 ### Fixed
-
+- Fixed issue where `NetworkSceneManager` did not synchronize despawned in-scene placed NetworkObjects. (#1898)
```
this section is now under pre.8 tag

0xFA11 commented on com.unity.netcode.gameobjects/Runtime/Core/NetworkObject.cs:
```diff
@@ -16,6 +16,11 @@ public sealed class NetworkObject : MonoBehaviour
         [SerializeField]
         internal uint GlobalObjectIdHash;
 
+        /// <summary>
+        /// For in-scene placed NetworkObjects, this is used to uniquely identify the
+        /// NetworkObject relative to the NetworkSceneHandle.
+        /// </summary>
+        internal int NetworkSceneHandle;
```
this doesn't seem like an innocent trivial change to me 👀

NoelStephensUnity replied:
I think I can remove that... glad you called it out.

0xFA11 commented on com.unity.netcode.gameobjects/Runtime/Core/NetworkObject.cs:
```diff
@@ -862,16 +868,17 @@ public struct TransformData : INetworkSerializeByMemcpy
             public NetworkObject OwnerObject;
             public ulong TargetClientId;
 
+            public int NetworkSceneHandle;
```
why we have `NetworkSceneHandle` here and also in `HeaderData` above?

NoelStephensUnity replied:
Because the one in the header was a very early implementation and was not supposed to be there... >.<
Removing the one from the header.

NoelStephensUnity committed:
test - fix

Removing m_CanCheckLoadUnloadEvents as that was a sanity check while I was testing clients loading scenes.  This could possibly fix an edge case scenario issue where a message might get ignored.

NoelStephensUnity committed:
update

removing NetworkSceneHandle from NetworkObject as it was not needed.

NoelStephensUnity committed:
Update CHANGELOG.md

NoelStephensUnity committed:
update

Removing wait for clients to unload scenes and unloading them via SceneManager as the test has already completed and this is more clean up than anything else.

NoelStephensUnity committed:
test update

Removed the crazy code in ObjectNameIdentifier for adding tag information and simplified it by storing the original name and using that each time it needs to regenerate the tag information.
Moved where the NetworkManager name and tag information is applied to get the right client id.
Added additional debug information.

NoelStephensUnity committed:
style

NoelStephensUnity committed:
test

Testing with verbose mode to see why this passes on Windows 10 in a stand alone build and in the editor but fails on Yamato desktops.

NoelStephensUnity committed:
test

adding more debug info.

NoelStephensUnity committed:
test fix

Looks like the scene isn't really loaded when it says it is loaded.
Moving when the NetworkSceneManager's callback is invoked and adding delay before invoking.

NoelStephensUnity committed:
test fix

Making some additional modifications to see if it was when the NetworkSceneManager scene Load/Unload event action is invoked and reducing the number of clients to 1.

NoelStephensUnity committed:
Merge branch 'develop' into fix/mtt-2924-despawned-scene-objects-dont-synchronize-late-joining-clients

NoelStephensUnity committed:
Test Fix

Minor tweaks and merging develop into this branch.

NoelStephensUnity committed:
test fix

Found a bug in some recent changes that would cause a crash under specific scenarios.

NoelStephensUnity committed:
fix

NoelStephensUnity committed:
test and style

Went back through some of the changed files and removed some code no longer needed, extended some of the verbose debug info for any future debugging purposes, and updated some of the comments.

NoelStephensUnity committed:
style

removing commented out code.

NoelStephensUnity committed:
Merge branch 'develop' into fix/mtt-2924-despawned-scene-objects-dont-synchronize-late-joining-clients

NoelStephensUnity committed:
Merge branch 'develop' into fix/mtt-2924-despawned-scene-objects-dont-synchronize-late-joining-clients

JesseOlmer commented:
Closed and re-opened to fix the JIRA association. Please make sure to add the JIRA ticket to the title upon PR creation in the future.

NoelStephensUnity committed:
Merge branch 'develop' into fix/mtt-2924-despawned-scene-objects-dont-synchronize-late-joining-clients

NoelStephensUnity committed:
Merge branch 'develop' into fix/mtt-2924-despawned-scene-objects-dont-synchronize-late-joining-clients

NoelStephensUnity committed:
Merge branch 'develop' into fix/mtt-2924-despawned-scene-objects-dont-synchronize-late-joining-clients
